File: java/scf-simple/src/main/java/org/mydotey/scf/source/stringproperty/memorymap/MemoryMapConfigurationSource.java
package org.mydotey.scf.source.stringproperty.memorymap;

import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

import org.mydotey.scf.ConfigurationSourceConfig;
import org.mydotey.scf.source.stringproperty.StringPropertyConfigurationSource;

/**
 * @author koqizhao
 *
 * May 17, 2018
 * 
 * all properties stored in a memory concurrent HashMap
 * 
 * dynamic source
 */
public class MemoryMapConfigurationSource extends StringPropertyConfigurationSource<ConfigurationSourceConfig> {

    private ConcurrentHashMap<String, String> _properties;

    public MemoryMapConfigurationSource(ConfigurationSourceConfig config) {
        super(config);

        _properties = new ConcurrentHashMap<>();
    }

    @Override
    public String getPropertyValue(String key) {
        return _properties.get(key);
    }

    public void setPropertyValue(String key, String value) {
        String oldValue = _properties.get(key);
        if (Objects.equals(oldValue, value))
            return;

        if (value == null)
            _properties.remove(key);
        else
            _properties.put(key, value);

        raiseChangeEvent();
    }

    public void setProperties(Map<String, String> properties) {
        _properties.putAll(properties);

        raiseChangeEvent();
    }

}

File: dotnet/src/MyDotey.SCF.Simple/Source/StringProperty/MemoryMap/MemoryMapConfigurationSource.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;

namespace MyDotey.SCF.Source.StringProperty.MemoryMap
{
    /**
     * @author koqizhao
     *
     * May 17, 2018
     * 
     * all properties stored in a memory concurrent Dictionary
     * 
     * dynamic source
     */
    public class MemoryMapConfigurationSource : StringPropertyConfigurationSource<ConfigurationSourceConfig>
    {
        private ConcurrentDictionary<string, string> _properties;

        public MemoryMapConfigurationSource(ConfigurationSourceConfig config)
            : base(config)
        {
            _properties = new ConcurrentDictionary<string, string>();
        }

        public override string getPropertyValue(string key)
        {
            _properties.TryGetValue(key, out string value);
            return value;
        }

        public virtual void setPropertyValue(string key, string value)
        {
            _properties.TryGetValue(key, out string oldValue);
            if (Object.Equals(oldValue, value))
                return;

            if (value == null)
                _properties.TryRemove(key, out value);
            else
                _properties[key] = value;

            raiseChangeEvent();
        }

        public virtual void setProperties(Dictionary<string, string> properties)
        {
            foreach (KeyValuePair<string, string> p in properties)
                _properties[p.Key] = p.Value;

            raiseChangeEvent();
        }
    }
}